h2o-core/src/main/java/hex/Model.java

```diff
@@ -604,6 +604,16 @@ public Frame removeInteractionOnlyColumns(Frame f) {
 
   }
 
+  public static class TEComponents extends Iced<TEComponents> {
+    private int[] _components;
```

Comment:
what are those? numerator and denominator?
wouldn't it deserve sth more explicit?
as well as `IcedHashMap<String, Map<String, TEComponents>>` imo: maps of maps are hard to reason about below as we don't see immediately what the keys stand for. Besides it would allow you to encapsulate logic like `fromFrames(Map<String, Frame> encodingMap)` or `fromMaps(Map.Entry<String, Map<String, int[]>)` making this testable + reusable anywhere

Reply:
@seb-h2o I totally agree that `numerator` and `denominator` are too broad empty terms, but do you have better suggestions? I would even call them something like 'countOfPositiveClassesInCategoricalLevel` and `sizeOfCategoricalLevel` but firstly... part `positive` would not fit when we add support for multinomial case and regression and secondly I wanted to be consistent with the naming in our first version of the TargetEncoding in R. For me it was making most sense giving the fact that in reality these two things are just components that help us compute frequencies for levels.

Reply:
@seb-h2o totally agree with the second part. That is why I had just Map<String, Frame> in TargetEncoder. But unfortunately there is no dependency to Frame `h2o-core` in `h2o-genmodel` package so sooner or later I would have to convert `Frame` into something. The best match is a Map I think. And as we don't know how many columns we are encoding I need to keep it generic so that is why I ended up with map of maps.
 I can probably wrap main Map into the class as I did with TEComponents to make it more clear what keys are stand for. 
Incapsulating it into a reusable method is something that I usually do when I'm sure that I understand that there is no other way to do something.... I was waiting from suggestions after review from @michalkurka about how to maybe avoid these conversions of the maps( maybe we have some helper methods to serialise Frames).

Reply:
@deil87 I don't mind `numerator/denominator`, these are relatively good terms for TE currently. It's `components` that I find too broad?
Is it in prevision for `multinomial` case?
> I can probably wrap main Map into the class as I did with TEComponents to make it more clear what keys are stand for.
That's what I mean

Reply:
@seb-h2o so you suggest to rename `TEComponents`? Agree... maybe `NumeratorDenominator` then? :) 
And let's wrap `Map<String, TEComponents>` into `EncodingMap`? wdyt?

Reply:
@deil87 I'm fine with `TEComponents`, it's about `_components`: do you plan to have more items in this array? Today it handles only numerator and numerator as an array, when everywhere else, you use 2 variables?

Reply:
@seb-h2o The main problem is that our `h2o-genmodel` module does not have any dependencies on other h2o modules. That is why I need to use native java datatypes to store what I need. The most reasonable structure in my opinion is `Map<String, Map<String, int[]>>`.  So even having `TEComponents` is unnecessary as we need to convert it in the end of the day to `Map<String, Map<String, int[]>>`.  So probably we need to improve readability of the code by adding comments and get rid of `TEComponents`.  wdyt?
 I'm using array to store numerator and denominator because I need to map categorical level to two components ( this part ->`Map<String, int[]>`). But what I can do is to rename `_components` into `_numeratorDenominator`.